# Exercise 5.5

In Chapter 4, we used logistic regression to predict the probability of `default` using `income` and `balance` on the `Default` data set. We will now estimate the test error of this logistic regression model using the validation set approach. Do not forget to set a random seed before beginning your analysis.

In [ ]:
set.seed(1)

## (a) Fit a logistic regression model that uses `income` and `balance` to predict `default`.

In [ ]:
glm.fit <- glm(default ~ income + balance, family="binomial", data=Default)
summary(glm.fit)


Call:
glm(formula = default ~ income + balance, family = "binomial", 
    data = Default)

Coefficients:
              Estimate Std. Error z value Pr(>|z|)    
(Intercept) -1.154e+01  4.348e-01 -26.545  < 2e-16 ***
income       2.081e-05  4.985e-06   4.174 2.99e-05 ***
balance      5.647e-03  2.274e-04  24.836  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 2920.6  on 9999  degrees of freedom
Residual deviance: 1579.0  on 9997  degrees of freedom
AIC: 1585

Number of Fisher Scoring iterations: 8


## (b) Using the validation set approach, estimate the test error of this model. In order to do this, you must perform the following steps:

i. Split the sample set into a training set and a validation set.

ii. Fit a multiple logistic regression model using only the training observations.

iii. Obtain a prediction of default status for each individual in the validation set by computing the posterior probability of default for that individual, and classifying the individual to the `default` category if the posterior probability is greater than 0.5.

iv. Compute the validation set error, which is the fraction of the observations in the validation set that are misclassified.

In [ ]:
train <- sample(10000, 5000)
glm.fit2 <- glm(default ~ income + balance, family="binomial", data=Default, subset=train)
glm.prob <- predict(glm.fit2, Default, type="response")
glm.pred <- rep("No", 10000)
glm.pred[glm.prob > .5] = "Yes"

table(glm.pred[-train], Default$default[-train])
error <- c(mean(glm.pred[-train] != Default$default[-train]))
error

     
        No  Yes
  No  4824  108
  Yes   19   49

[1] 0.0254

## (c) Repeat the process in (b) three times, using three different splits of the observations into a training set and a validation set. Comment on the results obtained.

In [ ]:
for (i in 1:3) {
    train <- sample(10000, 5000)
    glm.fit2 <- glm(default ~ income + balance, family="binomial", data=Default, subset=train)
    glm.prob <- predict(glm.fit2, Default, type="response")
    glm.pred <- rep("No", 10000)
    glm.pred[glm.prob > .5] = "Yes"

    table(glm.pred[-train], Default$default[-train])
    error <- c(error, mean(glm.pred[-train] != Default$default[-train]))
}
error
mean(error)

[1] 0.0254 0.0274 0.0244 0.0244

[1] 0.0254

The test error varies for each process, but the average error is 0.0254.

## (d) Now consider a logistic regression model that predicts the probability of `default` using `income`, `balance`, and a dummy variable for `student`. Estimate the test error for this model using the validation set approach. Comment on whether or not including a dummy variable for `student` leads to a reduction in the test error rate.

In [ ]:
train <- sample(10000, 5000)
glm.fit3 <- glm(default ~ ., family="binomial", data=Default, subset=train)
glm.prob <- predict(glm.fit3, Default, type="response")
glm.pred <- rep("No", 10000)
glm.pred[glm.prob > .5] = "Yes"

summary(glm.fit3)
table(glm.pred[-train], Default$default[-train])
mean(glm.pred[-train] != Default$default[-train])


Call:
glm(formula = default ~ ., family = "binomial", data = Default, 
    subset = train)

Coefficients:
              Estimate Std. Error z value Pr(>|z|)    
(Intercept) -1.042e+01  6.921e-01 -15.058   <2e-16 ***
studentYes  -8.286e-01  3.468e-01  -2.389   0.0169 *  
balance      5.638e-03  3.276e-04  17.212   <2e-16 ***
income      -5.472e-06  1.205e-05  -0.454   0.6498    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1409.45  on 4999  degrees of freedom
Residual deviance:  767.12  on 4996  degrees of freedom
AIC: 775.12

Number of Fisher Scoring iterations: 8


     
        No  Yes
  No  4808  121
  Yes   18   53

[1] 0.0278

A dummy variable for `student` seems statistically significant, but due to high correlation between predictors, the `income` variable now loses importance. The increased test error suggests that including a dummy variable for `student` does not lead to a reduction in the test error rate.